### PR TITLE
(2073) Handle user entered URLs not containing protocol

### PR DESCRIPTION
**Pull request by bryangaledxw**
We now silently add a default protocol if this is missing from the URL, and for any existing data in our database that is missing a URL protocol, we add the URL at presentation time.

### CODE REVIEW

File: src/professions/admin/dto/registration.dto.ts [hunk 1]
```diff
@@ -17,4 +17,6 @@ export class RegistrationDto {
   )
   @ValidateIf((e) => e.registrationUrl)
   registrationUrl: string;
+
+  change: string;
```
Comment (pezholio) [inline src/professions/admin/dto/registration.dto.ts:24]: Can we strongly type the contents of the string here? Maybe using a template literal?

File: src/organisations/admin/dto/organisation.dto.ts [hunk 2]
```diff
@@ -18,20 +23,16 @@ export class OrganisationDto {
...
-  @IsUrl(
-    {},
-    {
-      message: 'organisations.admin.form.errors.url.invalid',
-    },
-  )
+  @IsUrl(urlOptions, {
+    message: 'organisations.admin.form.errors.url.invalid',
+  })
+  @Transform(({ value }) => preprocessUrl(value))
```
Comment (pezholio) [inline src/organisations/admin/dto/organisation.dto.ts:29]: 👍

File: src/helpers/validator.ts [hunk 3]
```diff
@@ -1,19 +1,22 @@
...
   dtoType: any;
   errors: ValidationError[];
 
-  public static async validate(dtoType: any, obj: object) {
-    const object: object = plainToClass(dtoType, obj);
+  public static async validate<T extends object>(
+    dtoType: { new (): T },
+    obj: object,
+  ) {
+    const object: T = plainToClass(dtoType, obj);
```
Comment (pezholio) [inline src/helpers/validator.ts:13]: 👍

File: src/helpers/format-link.helper.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import { isURL } from 'class-validator';
+import { escape } from './escape.helper';
+import { preprocessUrl, urlOptions } from './preprocess-url.helper';
+
+export function formatLink(url: string, classes = ''): string {
```
Comment (pezholio) [inline src/helpers/format-link.helper.ts:5]: Should we have `govuk-link` as the default here?

File: views/admin/professions/check-your-answers.njk [hunk 5]
```diff
@@ -482,7 +482,7 @@
                     html: ("professions.form.label.legislation.link" | t) + "<span class=\"govuk-visually-hidden\"> " + loop.index + "</span>"
                   },
                   value: {
-                    html: ("<a class='govuk-link' href='" + (legislation.url | escape) + "'>" + (legislation.url | escape) + "</a>") if legislation else ""
+                    html: (legislation.url | link("govuk-link")) if legislation else ""
```
Comment (pezholio) [inline views/admin/professions/check-your-answers.njk:485]: Do we need this conditional here now?

**Review by pezholio — APPROVED** [review]
Lovely bit of work this - thanks! 👍 Just a couple of things, but nothing is a deal-breaker.

File: src/helpers/format-link.helper.ts (hunk 4 above)
Comment (bryangaledxw) [reply src/helpers/format-link.helper.ts:5]: Let's!

File: views/admin/professions/check-your-answers.njk (hunk 5 above)
Comment (bryangaledxw) [reply views/admin/professions/check-your-answers.njk]: We do not! I keep forgetting `.` in Nunjucks is basically `?.`

File: src/professions/admin/dto/registration.dto.ts (hunk 1 above)
Comment (bryangaledxw) [reply src/professions/admin/dto/registration.dto.ts]: I've changed these to:

```
@Transform(({ value }) => parseBoolean(value))
change: boolean;
```

**Pull request merged**